src/v/cloud_topics/frontend_reader/reader.cc

```diff
@@ -0,0 +1,166 @@
+  , _data_plane_api(data_plane_api)
+  , _stm_api(stm_api) {}
+
+bool cloud_topics_log_reader_impl::is_end_of_stream() const {
+    return _state == state::end_of_stream;
+}
+
+ss::future<model::record_batch_reader::storage_t>
+cloud_topics_log_reader_impl::do_load_slice(
+  model::timeout_clock::time_point deadline) {
+    // This simple loop and switch doesn't attempt to stitch L1 and L0 batches
+    // together. When reaching the end of L1, the reader will return whatever
+    // it gets and read from L0 on the next call to do_load_slice.
+    while (true) {
+        switch (_state) {
+        case state::end_of_stream:
+            co_return model::record_batch_reader::storage_t{};
+        case state::uninitialized:
+            co_await initialize_reader(deadline);
+            vassert(
+              _state != state::uninitialized, "reader should be initialized");
+            break;
+        case state::reading_l1:
+            co_return co_await read_l1(deadline);
+        case state::reading_l0:
+            co_return co_await read_l0(deadline);
+        }
+    }
+}
+
+void cloud_topics_log_reader_impl::print(std::ostream& o) {
+    o << "cloud_topics_log_reader";
+}
+
+ss::future<> cloud_topics_log_reader_impl::initialize_reader(
+  model::timeout_clock::time_point /*deadline*/) {
+    _lro = _stm_api->get_last_reconciled_offset();
+    _original_max_offset = _config.max_offset;
+
+    if (_config.start_offset <= _lro) {
+        vlog(
+          cd_log.debug,
+          "Initializing reader in L1 for {} with start_offset {} and LRO {}",
+          _ntp,
+          _config.start_offset,
+          _lro);
+
+        _config.max_offset = std::min(_config.max_offset, _lro);
+
+        _l1_reader = std::make_unique<level_one_log_reader_impl>(
+          _config, _ntp, _tidp, _metastore.get(), _io);
+        _state = state::reading_l1;
+        co_return;
+    }
+
+    vlog(
+      cd_log.debug,
+      "Initializing reader in L0 for {} with start_offset {} and LRO {}",
+      _ntp,
+      _config.start_offset,
+      _lro);
+
+    _l0_reader = std::make_unique<level_zero_log_reader_impl>(
+      _config, _partition, _data_plane_api);
+    _state = state::reading_l0;
+}
+
+bool cloud_topics_log_reader_impl::should_transition_to_l0() const {
+    if (_lro >= _original_max_offset) {
+        return false;
+    }
+
+    if (
+      (_config.strict_max_bytes || _config.bytes_consumed > 0)
+      && _config.bytes_consumed >= _config.max_bytes) {
+        return false;
+    }
+
+    return true;
+}
+
+ss::future<model::record_batch_reader::storage_t>
+cloud_topics_log_reader_impl::read_l1(
+  model::timeout_clock::time_point deadline) {
+    vassert(_state == state::reading_l1, "reading L1 state mismatch");
+    vassert(_l1_reader, "L1 reader should be initialized");
+    vassert(!_l0_reader, "L0 reader should not be initialized");
+
+    auto result = co_await _l1_reader->do_load_slice(deadline);
+
+    if (_l1_reader->is_end_of_stream()) {
```

Comment:
IMO, we should just consume the reader and when we reached EOS we should ask for another reader and continue reading from it. We shouldn't have a point at which we're switching from L1 to L0 because at this point we might get another L1 reader instead (because of races).